controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+			continue
+		}
+		// TODO: make this formation.Diff(actual)?
+		formation.Processes = actual
+		diff := formation.Update(expected)
+		formationLog.Info("existing formation in incorrect state", "expected", expected, "actual", actual, "diff", diff)
+		s.sendDiffRequests(formation, diff)
+	}
+}
+
+func (s *Scheduler) HandleFormationChange(ef *ct.ExpandedFormation) {
+	var err error
+	defer func() {
+		s.sendEvent(NewEvent(EventTypeFormationChange, err, nil))
+	}()
+
+	log := fnLogger()
+	if ef == nil || ef.App == nil || ef.Release == nil {
+		log.Error("invalid formation")
+		return
+	}
+
+	log = log.New("app.id", ef.App.ID, "release.id", ef.Release.ID)
+	log.Info("handling formation change")
+	_, err = s.changeFormation(ef)
+	if err != nil {
+		log.Error("error handling formation change", "err", err)
+		return
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) HandleJobRequest(req *JobRequest) {
+	log := fnLogger("req.id", req.JobID, "req.type", req.RequestType)
+
+	if !s.isLeader {
+		log.Warn("ignoring job request as not service leader")
+		return
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.Error("error handling job request", "err", err)
+		}
+		s.sendEvent(NewEvent(EventTypeJobRequest, err, req))
+	}()
+
+	log.Info("handling job request")
+	switch req.RequestType {
+	case JobRequestTypeUp:
+		// startJob sets the HostID on the request if successful
+		// otherwise it remains blank
+		err = s.startJob(req)
+	case JobRequestTypeDown:
+		err = s.stopJob(req)
+	default:
+		err = fmt.Errorf("unknown job request type: %s", req.RequestType)
+	}
+}
+
+func (s *Scheduler) RunPutJobs() {
+	log := fnLogger()
+	log.Info("starting job persistence loop")
+	strategy := attempt.Strategy{Delay: 100 * time.Millisecond, Total: time.Minute}
+	for {
+		job, ok := <-s.putJobs
+		if !ok {
+			log.Info("stopping job persistence loop")
+			return
+		}
+		jobLog := log.New("job.id", job.ID, "job.state", job.State)
+		jobLog.Info("persisting job")
+		err := strategy.RunWithValidator(func() error {
+			return s.PutJob(job)
+		}, httphelper.IsRetryableError)
+		if err != nil {
+			jobLog.Error("error persisting job", "err", err)
+		}
+	}
+}
+
+func (s *Scheduler) HandleLeaderChange(isLeader bool) {
+	log := fnLogger()
+	s.isLeader = isLeader
+	if isLeader {
+		log.Info("handling leader promotion")
+		s.triggerRectify()
+	} else {
+		log.Info("handling leader demotion")
+		// TODO: stop job restart timers
+	}
+	s.sendEvent(NewEvent(EventTypeLeaderChange, nil, isLeader))
+}
+
+func (s *Scheduler) sendDiffRequests(f *Formation, diff Processes) {
+	log := fnLogger("app.id", f.App.ID, "release.id", f.Release.ID)
+	for typ, n := range diff {
+		if n > 0 {
+			log.Info(fmt.Sprintf("requesting %d new job(s) of type %s", n, typ))
+			for i := 0; i < n; i++ {
+				req := NewJobRequest(f, JobRequestTypeUp, typ, "", random.UUID())
+				req.state = JobStateRequesting
+				s.jobs.AddJob(req.Job)
+				s.HandleJobRequest(req)
```

Comment:
Given the request is being explicitly handled immediately (rather than being sent to `s.jobRequests`), is there any need for the request? Could this not just be something like `startJob(NewJob(...))`?